File: clients/android/NewsBlur/app/src/main/java/com/newsblur/domain/Classifier.java
package com.newsblur.domain;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import android.content.ContentValues;
import android.database.Cursor;

import com.google.gson.annotations.SerializedName;
import com.newsblur.database.DatabaseConstants;

public class Classifier implements Serializable {
	
	private static final long serialVersionUID = 8958319817246110753L;

	public static final int AUTHOR = 0, FEED = 1, TITLE = 2, TAG = 3;
	public static final int LIKE = 1, DISLIKE = -1, CLEAR_DISLIKE = 3, CLEAR_LIKE = 4;

    // these pre/post constants are used to build the 16 possible parameter names accepted by the API
    // when updating intel
    private static final String AUTHOR_POSTFIX = "author";
    private static final String FEED_POSTFIX = "feed";
    private static final String TITLE_POSTFIX = "title";
    private static final String TAG_POSTFIX = "tag";
    private static final String LIKE_PREFIX = "like_";
    private static final String DISLIKE_PREFIX = "dislike_";
    private static final String CLEAR_LIKE_PREFIX = "remove_like_";
    private static final String CLEAR_DISLIKE_PREFIX = "remove_dislike_";
	
	@SerializedName("authors")
	public HashMap<String, Integer> authors = new HashMap<String, Integer>();
	
	@SerializedName("titles")
	public HashMap<String, Integer> title = new HashMap<String, Integer>();
	
	@SerializedName("tags")
	public HashMap<String, Integer> tags = new HashMap<String, Integer>();
	
	@SerializedName("feeds")
	public HashMap<String, Integer> feeds = new HashMap<String, Integer>();

    // not vended by API, but all classifiers are received in the context of a feed, where this is set. needs to
    // be set manually when unfrozen.
    public String feedId;
	
    public ValueMultimap getAPITuples() {
        ValueMultimap values = new ValueMultimap();
        for (Map.Entry<String,Integer> entry : authors.entrySet()) {
            values.put(buildAPITupleKey(entry.getValue(), AUTHOR_POSTFIX), entry.getKey());
        }
        for (Map.Entry<String,Integer> entry : title.entrySet()) {
            values.put(buildAPITupleKey(entry.getValue(), TITLE_POSTFIX), entry.getKey());
        }
        for (Map.Entry<String,Integer> entry : tags.entrySet()) {
            values.put(buildAPITupleKey(entry.getValue(), TAG_POSTFIX), entry.getKey());
        }
        for (Map.Entry<String,Integer> entry : feeds.entrySet()) {
            values.put(buildAPITupleKey(entry.getValue(), FEED_POSTFIX), entry.getKey());
        }
        return values;
    }

    private String buildAPITupleKey(int action, String postfix) {
        switch (action) {
            case LIKE:
                return (LIKE_PREFIX + postfix);
            case DISLIKE:
                return (DISLIKE_PREFIX + postfix);
            case CLEAR_LIKE:
                return (CLEAR_LIKE_PREFIX + postfix);
            case CLEAR_DISLIKE:
                return (CLEAR_DISLIKE_PREFIX + postfix);
            default:
                throw new IllegalArgumentException("invalid classifier action type");
        }
    }
	
	public List<ContentValues> getContentValues() {
		List<ContentValues> valuesList = new ArrayList<ContentValues>();
		for (String key : authors.keySet()) {
			ContentValues authorValues = new ContentValues();
			authorValues.put(DatabaseConstants.CLASSIFIER_ID, feedId);
			authorValues.put(DatabaseConstants.CLASSIFIER_KEY, key);
			authorValues.put(DatabaseConstants.CLASSIFIER_TYPE, AUTHOR);
			authorValues.put(DatabaseConstants.CLASSIFIER_VALUE, authors.get(key));
			
			valuesList.add(authorValues);
		}
		
		for (String key : title.keySet()) {
			ContentValues titleValues = new ContentValues();
			titleValues.put(DatabaseConstants.CLASSIFIER_ID, feedId);
			titleValues.put(DatabaseConstants.CLASSIFIER_KEY, key);
			titleValues.put(DatabaseConstants.CLASSIFIER_TYPE, TITLE);
			titleValues.put(DatabaseConstants.CLASSIFIER_VALUE, title.get(key));
			
			valuesList.add(titleValues);
		}
		
		for (String key : tags.keySet()) {
			ContentValues tagValues = new ContentValues();
			tagValues.put(DatabaseConstants.CLASSIFIER_ID, feedId);
			tagValues.put(DatabaseConstants.CLASSIFIER_KEY, key);
			tagValues.put(DatabaseConstants.CLASSIFIER_TYPE, TAG);
			tagValues.put(DatabaseConstants.CLASSIFIER_VALUE, tags.get(key));
			
			valuesList.add(tagValues);
		}
		
		for (String key : feeds.keySet()) {
			ContentValues feedValues = new ContentValues();
			feedValues.put(DatabaseConstants.CLASSIFIER_ID, feedId);
			feedValues.put(DatabaseConstants.CLASSIFIER_KEY, key);
			feedValues.put(DatabaseConstants.CLASSIFIER_TYPE, FEED);
			feedValues.put(DatabaseConstants.CLASSIFIER_VALUE, feeds.get(key));
			
			valuesList.add(feedValues);
		}
		
		return valuesList;
	}
	
	public static Classifier fromCursor(final Cursor cursor) {
		Classifier classifier = new Classifier();
		
		while (cursor.moveToNext()) {
			String key = cursor.getString(cursor.getColumnIndex(DatabaseConstants.CLASSIFIER_KEY));
			int value = cursor.getInt(cursor.getColumnIndex(DatabaseConstants.CLASSIFIER_VALUE));
			
			switch (cursor.getInt(cursor.getColumnIndex(DatabaseConstants.CLASSIFIER_TYPE))) {
			case AUTHOR:
				classifier.authors.put(key, value);
				break;
			case TITLE:
				classifier.title.put(key, value);	
				break;
			case FEED:
				classifier.feeds.put(key, value);
				break;
			case TAG:
				classifier.tags.put(key, value);
				break;	
			}
		}
		
		return classifier;
	}
	
}

File: clients/android/NewsBlur/app/src/main/java/com/newsblur/domain/Classifier.kt
package com.newsblur.domain

import android.content.ContentValues
import android.database.Cursor
import com.google.gson.annotations.SerializedName
import com.newsblur.database.DatabaseConstants
import java.io.Serializable

class Classifier : Serializable {
    @JvmField
    @SerializedName("authors")
    val authors: MutableMap<String, Int> = mutableMapOf()

    @JvmField
    @SerializedName("titles")
    val title: MutableMap<String, Int> = mutableMapOf()

    @JvmField
    @SerializedName("tags")
    val tags: MutableMap<String, Int> = mutableMapOf()

    @JvmField
    @SerializedName("feeds")
    val feeds: MutableMap<String, Int> = mutableMapOf()

    @JvmField
    @SerializedName("texts")
    val texts: MutableMap<String, Int> = mutableMapOf()

    @JvmField
    @SerializedName("text_regex")
    val textRegex: MutableMap<String, Int> = mutableMapOf()

    @JvmField
    @SerializedName("urls")
    val urls: MutableMap<String, Int> = mutableMapOf()

    @JvmField
    @SerializedName("url_regex")
    val urlRegex: MutableMap<String, Int> = mutableMapOf()

    @JvmField
    @SerializedName("title_regex")
    val titleRegex: MutableMap<String, Int> = mutableMapOf()

    // there are scenarios where this is not vended by API; must be set manually
    @JvmField
    var feedId: String? = null

    fun hasStoryTextHighlights(): Boolean =
        texts.isNotEmpty() || textRegex.isNotEmpty()

    fun getAPITuples(): ValueMultimap {
        val values = ValueMultimap()

        authors.forEach { (key, value) ->
            buildAPITupleKey(value, AUTHOR_POSTFIX)?.let { values.put(it, key) }
        }
        title.forEach { (key, value) ->
            buildAPITupleKey(value, TITLE_POSTFIX)?.let { values.put(it, key) }
        }
        tags.forEach { (key, value) ->
            buildAPITupleKey(value, TAG_POSTFIX)?.let { values.put(it, key) }
        }
        feeds.forEach { (key, value) ->
            buildAPITupleKey(value, FEED_POSTFIX)?.let { values.put(it, key) }
        }
        texts.forEach { (key, value) ->
            buildAPITupleKey(value, TEXT_POSTFIX)?.let { values.put(it, key) }
        }
        urls.forEach { (key, value) ->
            buildAPITupleKey(value, URL_POSTFIX)?.let { values.put(it, key) }
        }

        return values
    }

    private fun buildAPITupleKey(
        action: Int,
        postfix: String,
    ): String? =
        when (action) {
            LIKE -> LIKE_PREFIX + postfix
            DISLIKE -> DISLIKE_PREFIX + postfix
            SUPER_DISLIKE -> SUPER_DISLIKE_PREFIX + postfix
            CLEAR_LIKE -> CLEAR_LIKE_PREFIX + postfix
            CLEAR_DISLIKE -> CLEAR_DISLIKE_PREFIX + postfix
            CLEAR_SUPER_DISLIKE -> CLEAR_SUPER_DISLIKE_PREFIX + postfix
            else -> null
        }

    fun getContentValues(): List<ContentValues> {
        val valuesList = mutableListOf<ContentValues>()

        authors.forEach { (key, value) ->
            val authorValues =
                ContentValues().apply {
                    put(DatabaseConstants.CLASSIFIER_ID, feedId)
                    put(DatabaseConstants.CLASSIFIER_KEY, key)
                    put(DatabaseConstants.CLASSIFIER_TYPE, AUTHOR)
                    put(DatabaseConstants.CLASSIFIER_VALUE, value)
                }
            valuesList.add(authorValues)
        }

        title.forEach { (key, value) ->
            val titleValues =
                ContentValues().apply {
                    put(DatabaseConstants.CLASSIFIER_ID, feedId)
                    put(DatabaseConstants.CLASSIFIER_KEY, key)
                    put(DatabaseConstants.CLASSIFIER_TYPE, TITLE)
                    put(DatabaseConstants.CLASSIFIER_VALUE, value)
                }
            valuesList.add(titleValues)
        }

        tags.forEach { (key, value) ->
            val tagValues =
                ContentValues().apply {
                    put(DatabaseConstants.CLASSIFIER_ID, feedId)
                    put(DatabaseConstants.CLASSIFIER_KEY, key)
                    put(DatabaseConstants.CLASSIFIER_TYPE, TAG)
                    put(DatabaseConstants.CLASSIFIER_VALUE, value)
                }
            valuesList.add(tagValues)
        }

        feeds.forEach { (key, value) ->
            val feedValues =
                ContentValues().apply {
                    put(DatabaseConstants.CLASSIFIER_ID, feedId)
                    put(DatabaseConstants.CLASSIFIER_KEY, key)
                    put(DatabaseConstants.CLASSIFIER_TYPE, FEED)
                    put(DatabaseConstants.CLASSIFIER_VALUE, value)
                }
            valuesList.add(feedValues)
        }

        texts.forEach { (key, value) ->
            val textValues =
                ContentValues().apply {
                    put(DatabaseConstants.CLASSIFIER_ID, feedId)
                    put(DatabaseConstants.CLASSIFIER_KEY, key)
                    put(DatabaseConstants.CLASSIFIER_TYPE, TEXT)
                    put(DatabaseConstants.CLASSIFIER_VALUE, value)
                }
            valuesList.add(textValues)
        }

        textRegex.forEach { (key, value) ->
            val regexValues =
                ContentValues().apply {
                    put(DatabaseConstants.CLASSIFIER_ID, feedId)
                    put(DatabaseConstants.CLASSIFIER_KEY, key)
                    put(DatabaseConstants.CLASSIFIER_TYPE, TEXT_REGEX)
                    put(DatabaseConstants.CLASSIFIER_VALUE, value)
                }
            valuesList.add(regexValues)
        }

        urls.forEach { (key, value) ->
            val urlValues =
                ContentValues().apply {
                    put(DatabaseConstants.CLASSIFIER_ID, feedId)
                    put(DatabaseConstants.CLASSIFIER_KEY, key)
                    put(DatabaseConstants.CLASSIFIER_TYPE, URL)
                    put(DatabaseConstants.CLASSIFIER_VALUE, value)
                }
            valuesList.add(urlValues)
        }

        urlRegex.forEach { (key, value) ->
            val regexValues =
                ContentValues().apply {
                    put(DatabaseConstants.CLASSIFIER_ID, feedId)
                    put(DatabaseConstants.CLASSIFIER_KEY, key)
                    put(DatabaseConstants.CLASSIFIER_TYPE, URL_REGEX)
                    put(DatabaseConstants.CLASSIFIER_VALUE, value)
                }
            valuesList.add(regexValues)
        }

        titleRegex.forEach { (key, value) ->
            val regexValues =
                ContentValues().apply {
                    put(DatabaseConstants.CLASSIFIER_ID, feedId)
                    put(DatabaseConstants.CLASSIFIER_KEY, key)
                    put(DatabaseConstants.CLASSIFIER_TYPE, TITLE_REGEX)
                    put(DatabaseConstants.CLASSIFIER_VALUE, value)
                }
            valuesList.add(regexValues)
        }

        return valuesList
    }

    companion object {
        const val AUTHOR: Int = 0
        const val FEED: Int = 1
        const val TITLE: Int = 2
        const val TAG: Int = 3
        const val TEXT: Int = 4
        const val TEXT_REGEX: Int = 5
        const val URL: Int = 6
        const val URL_REGEX: Int = 7
        const val TITLE_REGEX: Int = 8

        const val LIKE: Int = 1
        const val DISLIKE: Int = -1
        const val SUPER_DISLIKE: Int = -2
        const val CLEAR_DISLIKE: Int = 3
        const val CLEAR_LIKE: Int = 4
        const val CLEAR_SUPER_DISLIKE: Int = 5

        // API key postfix/prefix constants
        private const val AUTHOR_POSTFIX = "author"
        private const val FEED_POSTFIX = "feed"
        private const val TITLE_POSTFIX = "title"
        private const val TAG_POSTFIX = "tag"
        private const val TEXT_POSTFIX = "text"
        private const val URL_POSTFIX = "url"
        private const val LIKE_PREFIX = "like_"
        private const val DISLIKE_PREFIX = "dislike_"
        private const val SUPER_DISLIKE_PREFIX = "super_dislike_"
        private const val CLEAR_LIKE_PREFIX = "remove_like_"
        private const val CLEAR_DISLIKE_PREFIX = "remove_dislike_"
        private const val CLEAR_SUPER_DISLIKE_PREFIX = "remove_super_dislike_"

        @JvmStatic
        fun fromCursor(cursor: Cursor): Classifier {
            val classifier = Classifier()

            val keyIndex = cursor.getColumnIndexOrThrow(DatabaseConstants.CLASSIFIER_KEY)
            val valueIndex = cursor.getColumnIndexOrThrow(DatabaseConstants.CLASSIFIER_VALUE)
            val typeIndex = cursor.getColumnIndexOrThrow(DatabaseConstants.CLASSIFIER_TYPE)

            while (cursor.moveToNext()) {
                val key = cursor.getString(keyIndex)
                val value = cursor.getInt(valueIndex)

                when (cursor.getInt(typeIndex)) {
                    AUTHOR -> classifier.authors[key] = value
                    TITLE -> classifier.title[key] = value
                    FEED -> classifier.feeds[key] = value
                    TAG -> classifier.tags[key] = value
                    TEXT -> classifier.texts[key] = value
                    TEXT_REGEX -> classifier.textRegex[key] = value
                    URL -> classifier.urls[key] = value
                    URL_REGEX -> classifier.urlRegex[key] = value
                    TITLE_REGEX -> classifier.titleRegex[key] = value
                }
            }

            return classifier
        }
    }
}
